testground/benchmark/benchmark/main.py

```diff
@@ -20,20 +24,47 @@ def entrypoint(ctx: Context):
     cli = ChainCommand(CRONOSD_PATH)
 
     # build the genesis file collectively, and setup the network topology
-    peer = bootstrap(ctx, cli)
+    bootstrap(ctx, cli)
 
     # start the node
-    kwargs = {"stdout": subprocess.DEVNULL}
-    if ctx.is_leader:
-        del kwargs["stdout"]
-    proc = subprocess.Popen([CRONOSD_PATH, "start"], **kwargs)
+    logfile = Path(ctx.params.test_outputs_path) / "node.log"
+    proc = subprocess.Popen(
+        [CRONOSD_PATH, "start"],
+        stdout=open(logfile, "ab", buffering=0),
+    )
 
     wait_for_port(26657)
     wait_for_port(8545)
     wait_for_block(cli, 1)
 
+    test_finish_entry = f"finish-test-{ctx.params.test_group_id}"
     if not ctx.is_validator:
-        sendtx(cli, peer)
+        w3 = web3.Web3(web3.providers.HTTPProvider("http://localhost:8545"))
+        assert w3.eth.chain_id == 777
+        genesis_account = export_eth_account(cli, "account")
+        accounts = fund_test_accounts(w3, genesis_account, ctx.params.num_accounts)
+        with ThreadPoolExecutor(max_workers=ctx.params.num_accounts) as executor:
+            futs = (
+                executor.submit(sendtx, w3, acct, ctx.params.num_txs)
+                for acct in accounts
+            )
+            for fut in as_completed(futs):
+                try:
+                    fut.result()
+                except Exception as e:
+                    print("test task failed", e)
```

Comment:
do we need stop early if fail?

Reply:
will try to run end, in the end, we calculate the processed tps.